### PR TITLE
feat(cognitoidentityprovider): Add support for Cognito User-Pools User to Group membership relations

### LINKED ISSUE #1383: Support for Cognito User-Pools User to Group membership relations
### What resource do you need?

A Cognito User-Pool Group to User assignment is required. This is not mapped by the Amazon Cognito User-Pools API and therefore represents some sort of _virtual_ resource.

### What is your use case?

User-to-Group Memberships should be declaratively represented by a `CR` in order to manage User-to-Group assignments.

### Would you be willing to contribute it using [code generator](https://github.com/crossplane/provider-aws/blob/master/CODE_GENERATION.md)?

In case the code generator supports this: yes. Initial approach on utilizing the code generator failed due to insufficient features.

### PR DESCRIPTION
**Pull request by abacus3**
### Description of your changes

Closes #1383

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

[contribution process]: https://git.io/fj2m9

- Automated unit tests have been written for basic controller functionality.
- Manually tested following test cases:
    - Username existing, Usergroup existing: k apply, k get, k delete
    - Username non-existing, Usergroup existing: k apply, k get, k delete
    - Username non-existing, Usergroup existing: k apply, k get, k delete
    - Username non-existing, Usergroup non-existing: k apply, k get, k delete

### CODE REVIEW

File: apis/cognitoidentityprovider/manualv1alpha1/doc.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package manualv1alpha1 contains managed resources for AWS identity services such as
+// IAM.
```
Comment (haarchri) [inline apis/cognitoidentityprovider/manualv1alpha1/doc.go:18]: Nitpick: please change the comment

File: apis/cognitoidentityprovider/manualv1alpha1/groupusermembership_types.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+type GroupUserMembershipParameters struct {
+	// Region is which region the Group will be created.
+	// +kubebuilder:validation:Required
+	Region string `json:"region"`
+
+	// UserPoolID is the Amazon Cognito Group Name (Group) of the Cognito User-Pool group you want to
+	// add User to.
+	// +immutable
+	// +crossplane:generate:reference:type=github.com/crossplane-contrib/provider-aws/apis/cognitoidentityprovider/v1alpha1.UserPool
+	// +crossplane:generate:reference:refFieldName=UserPoolIDRef
```
Comment (haarchri) [inline apis/cognitoidentityprovider/manualv1alpha1/groupusermembership_types.go:42]: You don't need this - Code Generator will pick it

File: apis/cognitoidentityprovider/manualv1alpha1/groupusermembership_types.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	// +crossplane:generate:reference:selectorFieldName=UserPoolIDSelector
```
Comment (haarchri) [inline apis/cognitoidentityprovider/manualv1alpha1/groupusermembership_types.go:43]: You don't need this - Code Generator will pick it

File: apis/cognitoidentityprovider/manualv1alpha1/groupusermembership_types.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+	// UserPoolIDSelector selects a reference to an Group to retrieve its userPoolId
+	// +optional
+	UserPoolIDSelector *xpv1.Selector `json:"userPoolIdSelector,omitempty"`
+
+	// Groupname is the Amazon Cognito Group Name (Group) of the Cognito User-Pool group you want to
+	// add User to.
+	// +immutable
+	// +crossplane:generate:reference:type=github.com/crossplane-contrib/provider-aws/apis/cognitoidentityprovider/v1alpha1.Group
+	// +crossplane:generate:reference:refFieldName=GroupnameRef
```
Comment (haarchri) [inline apis/cognitoidentityprovider/manualv1alpha1/groupusermembership_types.go:59]: You don't need this - Code Generator will pick it

File: apis/cognitoidentityprovider/manualv1alpha1/groupusermembership_types.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	// +crossplane:generate:reference:selectorFieldName=GroupnameSelector
```
Comment (haarchri) [inline apis/cognitoidentityprovider/manualv1alpha1/groupusermembership_types.go:60]: You don't need this - Code Generator will pick it

File: examples/cognito/groupusermembership.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+apiVersion: cognitoidentityprovider.aws.crossplane.io/v1alpha1
+kind: GroupUserMembership
+metadata:
+  name: example-group-user-membership
+spec:
+  forProvider:
+    region: us-east-1
+    userPoolId: us-east-1_XXXXXXXX
```
Comment (haarchri) [inline examples/cognito/groupusermembership.yaml:8]: Can we use ref or selector in example ?

**Review by haarchri — COMMENTED** [review]
Please check a few small nitpicks then we can go

File: examples/cognito/groupusermembership.yaml (hunk 6 above)
Comment (abacus3) [reply examples/cognito/groupusermembership.yaml]: I've updated the example to use `userPoolIdRef` and `groupnameRef`.

File: apis/cognitoidentityprovider/manualv1alpha1/groupusermembership_types.go (hunk 5 above)
Comment (abacus3) [reply apis/cognitoidentityprovider/manualv1alpha1/groupusermembership_types.go]: Removed, please review

File: apis/cognitoidentityprovider/manualv1alpha1/groupusermembership_types.go (hunk 4 above)
Comment (abacus3) [reply apis/cognitoidentityprovider/manualv1alpha1/groupusermembership_types.go]: Removed, please review

File: apis/cognitoidentityprovider/manualv1alpha1/groupusermembership_types.go (hunk 3 above)
Comment (abacus3) [reply apis/cognitoidentityprovider/manualv1alpha1/groupusermembership_types.go]: Removed, please review

File: apis/cognitoidentityprovider/manualv1alpha1/groupusermembership_types.go (hunk 2 above)
Comment (abacus3) [reply apis/cognitoidentityprovider/manualv1alpha1/groupusermembership_types.go]: Removed, please review

File: apis/cognitoidentityprovider/manualv1alpha1/doc.go (hunk 1 above)
Comment (abacus3) [reply apis/cognitoidentityprovider/manualv1alpha1/doc.go]: Adjusted the comment, please review

**Review by haarchri — APPROVED** [review]
LGTM

**Pull request merged**